Do not use colors for logs when clowder is active

lzap opened the pull request:
SSIA

Looks ugly, e.g. https://ci.int.devshift.net/job/RHEnVision-provisioning-backend-pr-check/464/artifact/artifacts/k8s_artifacts/ephemeral-ewuczi/logs/provisioning-backend-api-service-7d8b9d888c-kn7tc_provisioning-backend-api-service.log

However, I don't know how to test this because our CI does not archive logs when job is successful. Any ideas @RHEnVision/qe ? :-)

akhil-jha commented:
hmm.....I see https://github.com/RHEnVision/provisioning-backend/pull/254 is failing in CI. I think combining the PRs would help. If  that's okay? :thinking:

lzap commented:
You know what, let me fail a test once to check logs, then I will rebase and fix this.

Do not merge, it will fail after the rebase :-)

lzap commented:
Will return back to this once https://github.com/RHEnVision/provisioning-backend/pull/258 is merged so I can use the new function `IsClowder`.

lzap commented:
It works! No color codes in https://ci.int.devshift.net/job/RHEnVision-provisioning-backend-pr-check/472/artifact/artifacts/k8s_artifacts/ephemeral-90nkar/logs/provisioning-backend-service-d9b858dc7-blm26_provisioning-backend-service.log

Ready for review. Updated commit message and rebased to follow the new commit message formatting guideline: https://github.com/RHEnVision/provisioning-backend/pull/261

akhil-jha commented:
/retest

lzap commented:
Rebased to solve conflicts.

ezr-ondrej commented on internal/logging/zerolog.go:
```diff
@@ -57,13 +56,22 @@ func decorate(l zerolog.Logger) zerolog.Logger {
+	var noColor bool
+
+	// Do not use colors when running in clowder
+	if config.InClowder() {
```
```suggestion
	if config.InEphemeralClowder() {
```

lzap replied:
This is not a typo, I introduced the function `InEphemeralClowder` but the intention was not to use it.

ezr-ondrej replied:
Ah, oki ;) i did not read well then :)

ezr-ondrej reviewed (commented):
Typo I guess :)

akhil-jha commented:
The test failing is expected here right? :thinking:

ezr-ondrej commented:
I didn't figure from logs, but I'd blame it on the typo in code.

lzap commented:
This patch should not be reason why pubkeys fail, it was green before the rebase.

Also, see that it actually works - no color ANSII codes anymore: https://ci.int.devshift.net/job/RHEnVision-provisioning-backend-pr-check/516/artifact/artifacts/k8s_artifacts/ephemeral-uypceo/logs/provisioning-backend-api-7b5d5d7bc9-nth29_provisioning-backend-api.log

lzap commented:
Looks like @ezr-ondrej is busy, can you review @adiabramovitch or @avitova please? I think this is ready. Rebased. The test will likely fail because there is some problem with instance termination in IQE.

Which makes it actually easier to check if this patch works - if you open up logs from our app, there should be no color codes. Here is an example of how these ugly color codes look like when dumped into a logfile:

```
[90m9:52AM[0m [32mINF[0m Worker starting [36mhostname=[0mprovisioning-backend-worker-868894c849-tx48g [36mversion=[0m1c035f5 [36mworker_id=[0mcd8hku5ctdpr9r2ea7l0
[90m9:52AM[0m [32mINF[0m Initializing redis application cache [36mhostname=[0mprovisioning-backend-worker-868894c849-tx48g [36mversion=[0m1c035f5
```

avitova commented on internal/logging/zerolog.go:
```diff
@@ -57,13 +56,22 @@ func decorate(l zerolog.Logger) zerolog.Logger {
+	var noColor bool
+
+	// Do not use colors when running in clowder
+	if config.InClowder() {
```
I know that noColor is false by default, but we are calling the config.InClowder() anyways.
Could we (1) possibly just initialize the noColor with the `config.InClowder()`? The return value is always bool anyways... Or do you see a problem with readability? I do not want this to block the PR, and I know it is not a problem in Go.

lzap replied:
Haha yeah of course, this happens when you end up refactoring over and over again until you left with such a non-sense. Good point.

avitova reviewed (approved):
Looks good and works for me. I just have a small nitpick, but if the thing I commented is justified, let's merge this. :)

akhil-jha commented:
/retest

lzap commented:
Rebased, incorporated @avitova 's recommendation.

ezr-ondrej commented on deploy/clowdapp.yaml:
```diff
@@ -168,6 +184,18 @@ parameters:
     value: quay.io/cloudservices/provisioning-backend
+  - description: Logging level (trace, debug, info, warn, error, fatal, panic)
+    name: LOGGING_LEVEL
+    value: "trace"
+  - description: Postgres driver logging level (trace, debug, info, warn, error, fatal, panic)
+    name: DATABASE_LOGGING_LEVEL
+    value: "trace"
```
This will enable trace level for stage as well... I believe we should set these to "resonable" values and set the different values per environment in app-interface

lzap replied:
Yeah but I don't know how to do that now, so I was hoping we can live with increased verbosity for some time.

ezr-ondrej replied:
Oki, I'm fine with that. If we can set no colors only in test env I'm ok. Or i can set levels to traces by parameters for ephemeral only if we agree to it

lzap replied:
So it looks like I figured out how to set ephemeral variables (through app-interface).

ezr-ondrej commented on internal/logging/zerolog.go:
```diff
@@ -62,8 +61,11 @@ func InitializeStdout() {
-		Out:        os.Stdout,
+		Out: os.Stdout,
+		// Do not use colors when running in clowder
+		NoColor:    config.InClowder(),
```
Are we sure we want to disable colors in stage env? the openshift logging shows the colors pretty well and it improves readability IMHO

lzap replied:
There is no terminal you can connect to, therefore it shows up unreadable. See the example here: https://github.com/RHEnVision/provisioning-backend/pull/255#issuecomment-1285484097

ezr-ondrej replied:
Well, there is none in the tests, but there is one in the stage environment namespace.
So I'd love to distinguish those two environments.

lzap replied:
It`s now info by default.

ezr-ondrej reviewed (commented):
I'm happy with these changes for ephemeral, but I'd not want to have them in stage, or is that intentional?

lzap commented:
I will look on how to create different logging config for stage/prod.

lzap commented:
Do not merge, I screwed up with git.

lzap commented:
Ok finally it's correct and ready, @adiabramovitch is working on getting the two variables LOGGING_LEVEL and DATABASE_LOGGING_LEVEL set accordingly:

Stage:

- info
- info

Ephemeral:

- trace
- trace
